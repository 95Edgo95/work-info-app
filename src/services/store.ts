import createSagaMiddleware, {SagaMiddleware} from "redux-saga";
import {applyMiddleware, compose, createStore} from "redux";
import reducer from "services/reducer";
import saga from "services/saga";
import {fromJS} from "immutable";

let store: any;

const initialState: any = fromJS({});

const sagaMiddleware: SagaMiddleware<any> = createSagaMiddleware();

if (process.env && process.env.NODE_ENV !== "production") {
    const composeEnhancers: any = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    store = createStore(reducer, initialState, composeEnhancers(applyMiddleware(sagaMiddleware)));
} else {
    store = createStore(reducer, initialState, compose(applyMiddleware(sagaMiddleware)));
}

sagaMiddleware.run(saga);

export default store;
