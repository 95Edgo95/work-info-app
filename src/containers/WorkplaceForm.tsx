import {
  attemptCreateWorkplace,
  attemptGetWorkplace,
  attemptUpdateWorkplace
} from "store/workplace/WorkplaceActionCreators";
import {mapDispatchToProps, mapStateToProps} from "helpers/ReduxHelper";
import {IWorkplace} from "store/workplace/WorkplaceReducer";
import {ISelectorVariables} from "services/selector";
import {RouteComponentProps} from "react-router";
import {connect} from "react-redux";
import Form from "components/Form";
import * as React from "react";
import {Map} from "immutable";

const inputs: any[] = [
  {
    placeholder: "Country",
    defaultValue: "",
    name: "country",
  },
  {
    placeholder: "City",
    defaultValue: "",
    name: "city",
  },
  {
    placeholder: "Company",
    defaultValue: "",
    name: "company",
  },
  {
    placeholder: "Start Date",
    defaultValue: "",
    name: "startDate",
    type: "date",
  },
  {
    placeholder: "End Date",
    defaultValue: "",
    name: "endDate",
    type: "date",
  },
];

interface IWorkplaceFormReduxProps {
  workplaces: Map<string, IWorkplace>;
}

interface IWorkplaceFormDispatchProps {
  attemptUpdateWorkplace: (data: any, id: string) => void;
  attemptCreateWorkplace: (data: any) => void;
  attemptGetWorkplace: (id: string) => void;
}

interface IWorkplaceFormProps extends RouteComponentProps<{ id: string; workbookId: string; }>, IWorkplaceFormReduxProps, IWorkplaceFormDispatchProps {}

function WorkplaceForm(props: IWorkplaceFormProps) {
  const {
    match: {params: {id, workbookId}},
    attemptCreateWorkplace,
    attemptUpdateWorkplace,
    attemptGetWorkplace,
    workplaces,
    history,
  } = props;

  const workplace: any = id ? workplaces.get(id) : Map({
    country: "",
    city: "",
    company: "",
    startDate: "",
    endDate: "",
  });

  const formInputs = [
    ...inputs,
    {
      defaultValue: workbookId,
      name: "workbookId",
      type: "hidden",
    }
  ];

  const validations = {
    country: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z]+$/gm,
        message: "Only letters.",
      }
    ],
    city: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z]+$/gm,
        message: "Only letters.",
      }
    ],
    company: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "regex",
        option: /^[a-zA-Z\s]*$/gm,
        message: "Only letters and space!",
      }
    ],
    startDate: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "is-before",
        option: "endDate",
        message: "Start date must be before end date!",
      }
    ],
    endDate: [
      {
        type: "required",
        message: "Required!",
      },
      {
        type: "is-after",
        option: "startDate",
        message: "End date must be after start date!",
      }
    ],
  };

  React.useEffect(() => {
    if (id && !workplace) {
      attemptGetWorkplace(id);
    }
  }, []);

  if (!workplace) {
    return null;
  }

  return (
    <Form
      attemptUpdateEntity={attemptUpdateWorkplace}
      attemptCreateEntity={attemptCreateWorkplace}
      entitiesSize={workplaces.size}
      validations={validations}
      inputs={formInputs}
      entity={workplace}
      history={history}
    />
  );
}

const actions = {
  attemptCreateWorkplace,
  attemptUpdateWorkplace,
  attemptGetWorkplace
};

const selectorVariables: ISelectorVariables = {
  workplace: {
    workplaces: true,
  }
};

export default connect<IWorkplaceFormReduxProps, IWorkplaceFormDispatchProps, any>(mapStateToProps(selectorVariables), mapDispatchToProps(actions))(WorkplaceForm);
