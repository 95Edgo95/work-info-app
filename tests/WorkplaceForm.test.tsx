import WorkplaceForm from "../src/containers/WorkplaceForm";
import {BrowserRouter} from "react-router-dom";
import store from "../__mocks__/storeMock";
import {Provider} from "react-redux";
import { mount } from "enzyme";
import * as React from "react";

describe('WorkbookForm', () => {
  it('should render correctly', () => {
    const component = mount(
      <Provider store={store}>
        <BrowserRouter>
          <WorkplaceForm match={{ params: { id: "qwer", workbookId: "asdf" } }} />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
    component.unmount();
  });
});
