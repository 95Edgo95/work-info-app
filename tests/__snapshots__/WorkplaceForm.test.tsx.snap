// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkbookForm should render correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(WorkplaceForm)
        match={
          Object {
            "params": Object {
              "id": "qwer",
              "workbookId": "asdf",
            },
          }
        }
      >
        <WorkplaceForm
          attemptCreateWorkplace={[Function]}
          attemptGetWorkplace={[Function]}
          attemptUpdateWorkplace={[Function]}
          match={
            Object {
              "params": Object {
                "id": "qwer",
                "workbookId": "asdf",
              },
            }
          }
          workplaces={Immutable.Map {}}
        />
      </Connect(WorkplaceForm)>
    </Router>
  </BrowserRouter>
</Provider>
`;
